### PR TITLE
#464F-#465F-add-self-assignment-option-on-templatePermission

### PR DESCRIPTION
**Pull request by andreievg**
Required for front end: #464-#465-Add-assignments-logic

### Changes

* Add canSelfAssign to templatePermission.restrictions (and make sure generated reviewAssignments are set to that if it exists)
* Add new permissionName (reviewSelfAssignable - 10500)
* Add template permission and user permissionJoin of permissionName for template testReview, as per below diagram (`under reviewSelfAssignable-10500` branch)

Screen Shot 2021-03-25 at 12 37 42 PM

### CODE REVIEW

File: src/plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts [hunk 1]
```diff
@@ -45,6 +45,13 @@ module.exports['generateReviewAssignments'] = async function (input: any, DBConn
         ? restrictions?.templateSectionRestrictions
         : null
 
+      const canSelfAssign = restrictions?.canSelfAssign
+      // Automatic option for slef assignment if review is above level 1
+      const status =
+        canSelfAssign || nextReviewLevel > 1
+          ? AssignmentStatus.SELF_ASSIGN
+          : AssignmentStatus.AVAILABLE
+
```
Comment (CarlosNZ) [inline src/plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts:54]: This is going to conflict with what @nmadruga's done in #280, so you'll need to resolve that.

File: database/insertData/03_template_D_basicReviewTesting.js [hunk 2]
```diff
@@ -269,6 +269,7 @@ exports.queries = [
                 level: 1
                 restrictions: { templateSectionRestrictions: ["S2"] }
               }
+              { permissionNameId: 10500, restrictions: { canSelfAssign: true }, stageNumber:1, level: 1 }
```
Comment (CarlosNZ) [inline database/insertData/03_template_D_basicReviewTesting.js:272]: Just wondering why `canSelfAssign` needs to be a distinct restriction. Surely having review permission AND assign permissions for a given template means that user can self-assign? Or, if you want more fine-grained (i.e. can ONLY assign themselves, not someone else), wouldn't you just add a more specific "selfAssign" policy and connect it to a "selfAssign" permission name?

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Main thing is to resolve the conflicts with the updated master.

plus a couple of questions.

File: src/plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts (hunk 1 above)
Comment (andreievg) [reply src/plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts]: https://github.com/openmsupply/application-manager-server/pull/283/commits/acb17b21deb1bdee1565bb875ce888e8f8e49605

will need to restest

File: database/insertData/03_template_D_basicReviewTesting.js (hunk 2 above)
Comment (CarlosNZ) [inline database/insertData/03_template_D_basicReviewTesting.js:272]: Would you mind creating a new issue to migrate the `canSelfAssign` and `templateSectionRestrictions` to their own fields as we discussed, with a brief explanation about the "restrictions" field only being used for additional policy parameters.

**Review by CarlosNZ — APPROVED** [review]
Cool, just one request to make a new issue. 

Cheers.

File: database/insertData/03_template_D_basicReviewTesting.js (hunk 2 above)
Comment (nmadruga) [reply database/insertData/03_template_D_basicReviewTesting.js:272]: Created a placeholder note for this issue [here](https://github.com/orgs/openmsupply/projects/6#card-58400772)

**Pull request merged**